File: VVFocusedCollectionView/VVFocusedCollectionViewLayout.m
//
//  VVFocusedCollectionViewLayout.m
//  VVFocusedCollectionView
//
//  Created by Vivi Yang on 8/11/16.
//  Copyright © 2016 Vivi Yang. All rights reserved.
//

#import "VVFocusedCollectionViewLayout.h"

@interface VVFocusedCollectionViewLayout ()

@property (strong, nonatomic) NSMutableArray *arrayOfAttributes;

@end

@implementation VVFocusedCollectionViewLayout


#pragma mark - Required methods

/// Returns the width and height of the collection view’s contents. These values represent the width and height of all the content, not just the content that is currently visible. The collection view uses this information to configure its own content size for scrolling purposes.
- (CGSize)collectionViewContentSize {
    
    CGFloat precedingGap = self.smallCellSize.width / 2.0 + self.cellGap;
    
    NSInteger numOfItems = [self.collectionView numberOfItemsInSection:0];
    if (numOfItems == 0) {
        return CGSizeMake(0, CGRectGetHeight(self.collectionView.bounds));
        
    } else if (numOfItems == 1) {
        return CGSizeMake(precedingGap + self.largeCellSize.width, CGRectGetHeight(self.collectionView.bounds));
        
    } else {
        return CGSizeMake(precedingGap + self.largeCellSize.width + self.cellGap + self.smallCellSize.width + (self.cellGap + self.smallCellSize.width) * (numOfItems - 2), CGRectGetHeight(self.collectionView.bounds));
    }
}

/// Return layout information for all items whose view intersects the specified rectangle. Your implementation should return attributes for all visual elements, including cells, supplementary views, and decoration views.
- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect {
    
    NSMutableArray *arrayOfAttributes = [NSMutableArray array];
    
    // Loop through each section and see if rect intersects any header / cell
    
    // Loop through all cells
    for (int index = 0; index < [self.collectionView numberOfItemsInSection:0]; index++) {
        
        NSIndexPath *cellIndexPath = [NSIndexPath indexPathForItem:index inSection:0];
        UICollectionViewLayoutAttributes *cellAttributes = [self layoutAttributesForItemAtIndexPath:cellIndexPath];
        
        if (CGRectIntersectsRect(cellAttributes.frame, rect)) {
            [arrayOfAttributes addObject:cellAttributes];
        }
    }
    
    return arrayOfAttributes;
}

/// Return layout information for items in the collection view. You use this method to provide layout information only for items that have a corresponding cell. Do not use it for supplementary views or decoration views.
- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    return self.arrayOfAttributes[indexPath.item];
}

/// Return YES if the collection view requires a layout update or NO if the layout does not need to change. Subclasses can override it and return an appropriate value based on whether changes in the bounds of the collection view require changes to the layout of cells and supplementary views.
- (BOOL) shouldInvalidateLayoutForBoundsChange:(CGRect)newBound {
    return YES;
}


#pragma mark - Optional methods

/// Layout updates occur the first time the collection view presents its content and whenever the layout is invalidated explicitly or implicitly because of a change to the view. During each layout update, the collection view calls this method first to give your layout object a chance to prepare for the upcoming layout operation.
- (void)prepareLayout {
    [super prepareLayout];
    
    
    if (!self.arrayOfAttributes) {
        self.arrayOfAttributes = [NSMutableArray array];
    } else {
        [self.arrayOfAttributes removeAllObjects];
    }
    
    CGFloat precedingGap = self.smallCellSize.width / 2.0 + self.cellGap;
    
    static NSInteger focusedIndex = 0;
    
    CGFloat anchorX = self.collectionView.contentOffset.x + precedingGap;
    
    CGFloat distance, maxDistance, percentage;
    CGFloat maxDiffInSize = self.largeCellSize.width - self.smallCellSize.width;
    
    for (NSInteger i = 0; i < [self.collectionView numberOfItemsInSection:0]; i++) {
        UICollectionViewLayoutAttributes *attributes = [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:[NSIndexPath indexPathForItem:i inSection:0]];
        
        CGFloat x, y, w, h;
        
        x = precedingGap + i * (self.smallCellSize.width + self.cellGap);
        
        if (i == 0) {
            x = precedingGap;
        } else {
            x = CGRectGetMaxX([(UICollectionViewLayoutAttributes *)(self.arrayOfAttributes[i - 1]) frame]) + self.cellGap;
        }
        
        if (x == anchorX || (x < anchorX && x + self.smallCellSize.width + self.cellGap > anchorX)) {
            // This is the focused cell!
            focusedIndex = i;
            
        } else if (self.collectionView.contentOffset.x < 0) {
            focusedIndex = 0;
        }
        
        if (i == focusedIndex) {
            
            distance = fabs(anchorX - x);
            maxDistance = self.smallCellSize.width;
            percentage = MAX(0, MIN(1, 1 - distance / maxDistance));    // percentage must be between 0 and 1
            
            w = h = self.smallCellSize.width + percentage * maxDiffInSize;
            
        } else if (i == focusedIndex + 1) {
            
            distance = fabs(anchorX - x);
            maxDistance = self.largeCellSize.width + self.cellGap;
            percentage = MAX(0, MIN(1, 1 - distance / maxDistance));    // percentage must be between 0 and 1
            
            w = h = self.smallCellSize.width + percentage * maxDiffInSize;
            
        } else {
            w = h = self.smallCellSize.width;
        }
        
        y = -(h - self.smallCellSize.height) / 2.0;
        attributes.frame = CGRectMake(x, y, w, h);
        
        [self.arrayOfAttributes addObject:attributes];
    }
}

@end

File: VVFocusedCollectionView/VVFocusedCollectionViewLayout.swift
//
//  VVFocusedCollectionViewLayout.swift
//  VVFocusedCollectionView
//
//  Created by Vivi on 2020/6/20.
//  Copyright © 2020 Vivi Yang. All rights reserved.
//

import UIKit

@objcMembers
class VVFocusedCollectionViewLayout: UICollectionViewFlowLayout {
    var cellSizeObject = VVFocusedCollectionViewCellSizeObject()
    
    private var focusedIndex = 0
    private var arrayOfAttributes: [UICollectionViewLayoutAttributes] = []

    // MARK: - Required methods
    
    override var collectionViewContentSize: CGSize {
        let itemCount = collectionView?.numberOfItems(inSection: 0) ?? 0
        let collectionViewHeight = collectionView?.bounds.height ?? 0
        if itemCount == 0 {
            return CGSize(width: 0, height: collectionViewHeight)
        } else if itemCount == 1 {
            let width = cellSizeObject.firstCellPrecedingSpacing + cellSizeObject.largeCellSize.width
            return CGSize(width: width, height: collectionViewHeight)
        } else {
            let widthUpToFirstCell = cellSizeObject.firstCellPrecedingSpacing + cellSizeObject.largeCellSize.width
            let widthAfterFistCell = (cellSizeObject.cellSpacing + cellSizeObject.smallCellSize.width) * CGFloat(itemCount - 1)
            let width = widthUpToFirstCell + widthAfterFistCell
            return CGSize(width: width, height: collectionViewHeight)
        }
    }
    
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        var arrayOfAttributes: [UICollectionViewLayoutAttributes] = []
        
        // Loop through all cells and see if rect intersects any cell
        for index in 0 ..< (collectionView?.numberOfItems(inSection: 0) ?? 0) {
            let cellIndexPath = IndexPath(item: index, section: 0)
            if let cellAttributes = layoutAttributesForItem(at: cellIndexPath),
                rect.intersects(cellAttributes.frame) {
                arrayOfAttributes.append(cellAttributes)
            }
        }
        return arrayOfAttributes
    }
    
    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        guard indexPath.item < arrayOfAttributes.count else { return nil }
        return arrayOfAttributes[indexPath.item]
    }

    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
        return true
    }
    
    // MARK: - Optional methods
    
    override func prepare() {
        super.prepare()
        
        arrayOfAttributes.removeAll()
        
        let cellSizeRatio = cellSizeObject.largeCellSize.height / cellSizeObject.largeCellSize.width
        let contentOffsetX = collectionView?.contentOffset.x ?? 0
        
        // anchorX is at the min X of the focused cell when scrolling ended
        let anchorX = contentOffsetX + cellSizeObject.firstCellPrecedingSpacing
        
        var distance: CGFloat = 0
        var maxDistance: CGFloat = 0
        var percentage: CGFloat = 0
        
        let maxDiffInSize = cellSizeObject.largeCellSize.width - cellSizeObject.smallCellSize.width
        
        for index in 0 ..< (collectionView?.numberOfItems(inSection: 0) ?? 0) {
            let indexPath = IndexPath(item: index, section: 0)
            let attributes = UICollectionViewLayoutAttributes(forCellWith: indexPath)
            var x, y, w, h: CGFloat
            
            if index == 0 {
                x = cellSizeObject.firstCellPrecedingSpacing
            } else {
                let previousAttribute = arrayOfAttributes[index - 1]
                x = previousAttribute.frame.maxX + cellSizeObject.cellSpacing
            }
            
            let isFocusedCell =
                x == anchorX || (x < anchorX && x + cellSizeObject.smallCellSize.width + cellSizeObject.cellSpacing > anchorX)
            
            if isFocusedCell {
                focusedIndex = index
            } else if contentOffsetX < 0 {
                focusedIndex = 0
            }
            
            if index == focusedIndex {
                distance = abs(anchorX - x)
                maxDistance = cellSizeObject.smallCellSize.width
                percentage = max(0, min(1, 1 - distance / maxDistance))
                
                w = cellSizeObject.smallCellSize.width + percentage * maxDiffInSize
                h = w * cellSizeRatio
            } else if index == focusedIndex + 1 {
                distance = abs(anchorX - x)
                maxDistance = cellSizeObject.largeCellSize.width + cellSizeObject.cellSpacing
                percentage = max(0, min(1, 1 - distance / maxDistance))
                
                w = cellSizeObject.smallCellSize.width + percentage * maxDiffInSize
                h = w * cellSizeRatio
            } else {
                w = cellSizeObject.smallCellSize.width
                h = w * cellSizeRatio
            }
            
            y = (cellSizeObject.largeCellSize.height - h) / 2.0
            attributes.frame = CGRect(x: x, y: y, width: w, height: h)
            arrayOfAttributes.append(attributes)
        }
    }
}
